tests/zfs-tests/tests/functional/events/zed_slow_io.ksh

```diff
@@ -0,0 +1,191 @@
+# See the License for the specific language governing permissions
+# and limitations under the License.
+#
+# When distributing Covered Code, include this CDDL HEADER in each
+# file and include the License file at usr/src/OPENSOLARIS.LICENSE.
+# If applicable, add the following below this CDDL HEADER, with the
+# fields enclosed by brackets "[]" replaced with your own identifying
+# information: Portions Copyright [yyyy] [name of copyright owner]
+#
+# CDDL HEADER END
+#
+
+#
+# Copyright (c) 2023, Klara Inc.
+#
+
+# DESCRIPTION:
+#	Verify that vdev properties, slow_io_n and slow_io_t, work with ZED.
+#
+# STRATEGY:
+#	1. Create a pool with single vdev
+#	2. Set slow_io_n/slow_io_t to non-default values
+#	3. Inject slow io errors
+#	4. Verify that ZED degrades vdev
+#
+
+. $STF_SUITE/include/libtest.shlib
+
+TESTDIR="$TEST_BASE_DIR/zed_slow_io"
+VDEV="$TEST_BASE_DIR/vdevfile.$$"
+TESTPOOL="slow_io_pool"
+FILEPATH="$TESTDIR/slow_io.testfile"
+
+OLD_SLOW_IO=$(get_tunable ZIO_SLOW_IO_MS)
+OLD_SLOW_IO_EVENTS=$(get_tunable SLOW_IO_EVENTS_PER_SECOND)
+
+verify_runnable "both"
+
+function do_setup
+{
+	log_must truncate -s 1G $VDEV
+	default_setup_noexit $VDEV
+	log_must zpool events -c
+	log_must zfs set compression=off $TESTPOOL
+	log_must zfs set primarycache=none $TESTPOOL
+	log_must zfs set prefetch=none $TESTPOOL
+	log_must zfs set recordsize=512 $TESTPOOL
+	for i in {1..10}; do
+		log_must dd if=/dev/urandom of=${FILEPATH}$i bs=512 count=1
+	done
+	zpool sync
+}
+
+# intermediate cleanup
+function do_clean
+{
+	zpool status -s $TESTPOOL
+
+	log_must zinject -c all
+	log_must zpool destroy $TESTPOOL
+	log_must rm -f $VDEV
+}
+
+# final cleanup
+function cleanup
+{
+	log_must zinject -c all
+
+	# if pool still exists then something failed so log additional info
+	if poolexists $TESTPOOL ; then
+		log_note "$(zpool status -s $TESTPOOL)"
+		echo "=================== zed log search ==================="
+		grep "Diagnosis Engine" $ZEDLET_DIR/zed.log
+		destroy_pool $TESTPOOL
+	fi
+	log_must zed_stop
+
+	log_must rm -f $VDEV
+
+	log_must set_tunable64 ZIO_SLOW_IO_MS $OLD_SLOW_IO
+	log_must set_tunable64 SLOW_IO_EVENTS_PER_SECOND $OLD_SLOW_IO_EVENTS
+}
+
+function start_slow_io
+{
+	zpool sync
+	log_must set_tunable64 ZIO_SLOW_IO_MS 10
+	log_must set_tunable64 SLOW_IO_EVENTS_PER_SECOND 1000
+
+	log_must zinject -d $VDEV -D10:1 -T read $TESTPOOL
+	zpool sync
+}
+
+function stop_slow_io
+{
+	log_must set_tunable64 ZIO_SLOW_IO_MS $OLD_SLOW_IO
+	log_must set_tunable64 SLOW_IO_EVENTS_PER_SECOND $OLD_SLOW_IO_EVENTS
+
+	log_must zinject -c all
+}
+
+# Test default ZED settings:
+#    inject 10 events over 2.5 seconds, should not degrade.
+function default_degrade
+{
+	do_setup
+
+	start_slow_io
+	for i in {1..10}; do
+		dd if=${FILEPATH}$i of=/dev/null count=1 bs=512 2>/dev/null
+		sleep 0.25
+	done
+	stop_slow_io
+	zpool status -s $TESTPOOL
+
+	# give slow ZED a chance to process the delay events
+	sleep 18
+	zpool status -s $TESTPOOL
```

Comment:
should these be log_note `zpool status`?

Reply:
yes